### ISSUE TITLE
If there's an error reading the response body, it isn't handled

### ISSUE THREAD

**Issue by shanecav** [jmeas/fetch-dedupe#39]
I'm only using this library indirectly through `react-request` (hello again!), but I noticed an issue come up that I think comes from this library.

Here's my situation: The server I'm requesting from is currently having some issues (500 internal server error), and the response body is not valid JSON. So in the code in this lib that says `return res[responseTypeToUse]()` (currently on line 106 in src/index.js on master), an error is thrown. But because the following `then` doesn't have a rejection callback, that rejection is unhandled.

When the above happens, `react-request` actually seems to think the request is still "loading" forever, since the rejection is never handled, and it never exposes the response/error.

I think a fix could look something like this (again starting on line 106 of src/index.js on master):
```js
return res[responseTypeToUse]().then(
  data => {
    res.data = data;

    if (dedupe) {
      resolveRequest({ requestKey: requestKeyToUse, res });
    } else {
      return res;
    }
  },
  err => {
    res.data = null;

    if (dedupe) {
      resolveRequest({ requestKey: requestKeyToUse, res });
    } else {
      return res;
      // is this right? or should it return Promise.reject(err)?
      // seems like we'd still want to return the response object, but I'm not sure
    }
  }
);
```

I tested this out by hacking on node_modules like a pro, and it seemed to work in my case, but thought I'd check with you before creating a PR.

**Comment by jamesplease** [comment]
> hello again!

:wave: hey man!

> The server I'm requesting from is currently having some issues (500 internal server error), and the response body is not valid JSON.

Heh, we actually deal with this same problem with most of our backends at Netflix 😂 

We have a workaround that I tried to document, but apparently not well enough 🙈 . Check out [this question in the fetch-dedupe FAQ](https://github.com/jamesplease/fetch-dedupe#an-empty-response-body-is-throwing-an-error-what-gives). I also show an example of this in the react-request docs for [`responseType`](https://github.com/jamesplease/react-request#responsetype). That describes what we're doing to get around the problem, although we agree that the developer experience isn't super great here.

If you follow those guides, you should be able to get unblocked here without any changes to this lib. With that said, I like the solution you presented, and I think that is worth PR'ing :+1: !

@marlonpp @sprjr @tbranyen what do y'all think?

---

Even longer-term, I am thinking of dropping the `responseType` API altogether (at least from `fetch-dedupe`) to support streams more easily. There's a little more info on that in #31 , but progress has been slow 🐢 

---

Also cross-referencing #12 and #23 , which are related to this issue, as well as https://github.com/jamesplease/react-request/issues/110

---

**Update:** The code you have there looks good btw.

**Comment by shanecav** [comment]
🙈 Clearly I didn't search enough before creating this issue, sorry about that! And thanks for those links, they make perfect sense. Adding that `responseType` prop in react-request's Fetch should work for me for now, but I'll create that PR here, too. Thanks!

**Comment by jamesplease** [comment]
Resolved in #40 ; released as v3.0.0. Thanks for opening the issue @shanecav
